golang/cosmos/x/lien/lien.go

```diff
@@ -0,0 +1,130 @@
+	Locked      sdk.Int `json:"locked"`
+	Liened      sdk.Int `json:"liened"`
+}
+
+// NewPortHandler returns a port handler for the Keeper.
+func NewPortHandler(k Keeper) vm.PortHandler {
+	return portHandler{keeper: k}
+}
+
+const (
+	LIEN_GET_ACCOUNT_STATE = "LIEN_GET_ACCOUNT_STATE"
+	LIEN_SET_LIENED        = "LIEN_SET_LIENED"
+)
+
+// Receive implements the vm.PortHandler method.
+// Receives and processes a bridge message, returning the
+// JSON-encoded response or error.
+// See spec/02_messages.md for the messages and responses.
+func (ch portHandler) Receive(ctx *vm.ControllerContext, str string) (string, error) {
+	var msg portMessage
+	err := json.Unmarshal([]byte(str), &msg)
+	if err != nil {
+		return "", err
+	}
+	switch msg.Type {
+	case LIEN_GET_ACCOUNT_STATE:
+		return ch.handleGetAccountState(ctx.Context, msg)
+
+	case LIEN_SET_LIENED:
+		return ch.handleSetLiened(ctx.Context, msg)
+	}
+	return "", fmt.Errorf("unrecognized type %s", msg.Type)
+}
+
+// handleGetAccountState processes a LIEN_GET_ACCOUNT_STATE message.
+// See spec/02_messages.md for the messages and responses.
+func (ch portHandler) handleGetAccountState(ctx sdk.Context, msg portMessage) (string, error) {
+	addr, err := sdk.AccAddressFromBech32(msg.Address)
+	if err != nil {
+		return "", fmt.Errorf("cannot convert %s to address: %w", msg.Address, err)
+	}
+	denom := msg.Denom
+	if err = sdk.ValidateDenom(denom); err != nil {
+		return "", fmt.Errorf("invalid denom %s: %w", denom, err)
+	}
+	state := ch.keeper.GetAccountState(ctx, addr)
+	reply := msgAccountState{
+		CurrentTime: ctx.BlockTime().Unix(),
+		Total:       state.Total.AmountOf(denom),
+		Bonded:      state.Bonded.AmountOf(denom),
+		Unbonding:   state.Unbonding.AmountOf(denom),
+		Locked:      state.Locked.AmountOf(denom),
+		Liened:      state.Liened.AmountOf(denom),
+	}
+	bz, err := json.Marshal(&reply)
+	if err != nil {
+		return "", fmt.Errorf("cannot marshal %v: %w", reply, err)
+	}
+	return string(bz), nil
+}
+
+// handleSetLiened processes a LIEN_SET_LIENED message.
+// See spec/02_messages.md for the messages and responses.
+func (ch portHandler) handleSetLiened(ctx sdk.Context, msg portMessage) (string, error) {
+	addr, err := sdk.AccAddressFromBech32(msg.Address)
+	if err != nil {
+		return "", fmt.Errorf("cannot convert %s to address: %w", msg.Address, err)
+	}
+	denom := msg.Denom
+	if err = sdk.ValidateDenom(denom); err != nil {
+		return "", fmt.Errorf("invalid denom %s: %w", denom, err)
+	}
+	lien := ch.keeper.GetLien(ctx, addr)
+	oldAmount := lien.GetCoins().AmountOf(denom)
+	if msg.Amount.Equal(oldAmount) {
+		// no-op, no need to do anything
+		return "true", nil
+	} else if msg.Amount.LT(oldAmount) {
+		// always okay to reduce liened amount
+		diff := sdk.NewCoin(denom, oldAmount.Sub(msg.Amount))
+		lien.Coins = lien.GetCoins().Sub(sdk.NewCoins(diff))
+		ch.keeper.SetLien(ctx, addr, lien)
+		return "true", nil
+	} else {
+		// check if it's okay to increase lein
```

Comment:
Please only change this if you're forced to rebase.
```suggestion
		// check if it's okay to increase lien
```